### PR TITLE
remove catch-all exception handling in buildcache command

### LINKED ISSUE #15157: Catch errors and fail fast during buildcache creation.
> Just to be clear: I don't think my point above about catching the permission denied errors needs to be incorporated for this PR, necessarily.  Someone (probably me, since I wrote the code) can handle that in another PR.

_Originally posted by @opadron in https://github.com/spack/spack/pull/15090/review_comment/create_

### PR DESCRIPTION
**Pull request by opadron**
Closes #15157 
Issue originally identified in #15090

### CODE REVIEW

**Comment by gartung** [comment]
I think originally I had it die but someone wanted to option to continue creating buildcaches for other dependencies.

**Comment by opadron** [comment]
As far as other errors that can happen during build_tarball, I think "report the error and give up" is a sensible default.

I think if we want to do anything else, such as support a "best-effort" option, we really ought to enumerate the classes of Errors that Spack is willing to tolerate as well as the manner in which it should respond to those errors.

Off the top of my head, I can think of:

 1. Errors that prevent creation of the build cache, altogether.  There isn't anything more Spack can do under these conditions, and so there's no difference between fail-fast and best-effort policies.

Examples:
 - "permission denied" while trying to create the buildcache directory

 2. Scenarios where Spack fails to populate a build cache with a particular package, and the nature of the error is such that subsequent package failures are overwhelmingly likely.  This scenario isn't much different from the first, and I think failing fast is the reasonable response.

Examples:
 - Any failure due to a missing command that Spack depends on (e.g.: `tar`).  It's *conceivable* that an admin might install a usable version of `tar` just in time for Spack to successfully populate *some* of the later packages, but there's no practical reason to keep trying and hoping that something like that happens.  It would be more practical to just fail fast and let the user rerun the operation at a later time.

 3. Scenarios where Spack fails to populate a build cache with a particular package, but the error is likely to be limited to that package.  This is the only scenario where I think it makes sense to use a best-effort policy.  What I'm struggling with is actually coming up with examples of these errors.  

What was the original motivation for skipping?  Presumably, there had to have been errors we were running into and wanted to skip over?

**Comment by gartung** [comment]
I think is was something @obreitwi asked for but I don't remember the justification.

File: lib/spack/spack/cmd/buildcache.py [hunk 1]
```diff
@@ -361,13 +361,9 @@ def _createtarball(env, spec_yaml, packages, directory, key, no_deps, force,
...
-        try:
-            bindist.build_tarball(spec, outdir, force, rel,
-                                  unsigned, allow_root, signkey,
-                                  not no_rebuild_index)
-        except Exception as e:
-            tty.warn('%s' % e)
-            pass
+        bindist.build_tarball(spec, outdir, force, rel,
+                              unsigned, allow_root, signkey,
+                              not no_rebuild_index)
```
Comment (alalazo) [inline lib/spack/spack/cmd/buildcache.py]: `not no_rebuild_index` is definitely a nasty way to say `rebuild_index`

Comment (gartung) [reply lib/spack/spack/cmd/buildcache.py]: not my change 🤷‍♂

**Comment by obreitwi** [comment]
> I think is was something @obreitwi asked for but I don't remember the justification.

Unfortunately I do not remember asking for it, too, but I think it was related to #9440 where one error during tarball creation should not prevent the other packages from being archived.

Since even with multiprocessing we found the buildcache command to be not quite fast enough for archiving, we have since switched to a manual approach using plain old unix tools. So no regards from our side concerning the `buildcache` command. :+1:

**Comment by gartung** [comment]
@opadron This just needs a flake8 fix to be able to merge.

**Pull request merged**